modules/pam_env/pam_env.c

```diff
@@ -375,6 +353,215 @@ static int _assemble_line(FILE *f, char *buffer, int buf_len)
     return used;
 }
 
+static char **_read_file(const char*filename, const pam_handle_t *pamh)
+{
+    FILE *conf;
+    char buffer[BUF_SIZE];
+    char **lines = NULL;
+
+    D(("Env file name is: %s", filename));
+
+    if ((conf = fopen(filename,"r")) == NULL) {
+      pam_syslog(pamh, LOG_ERR, "Unable to open env file: %s", filename);
+      return NULL;
+    }
+
+    size_t i = 0;
+    while (_assemble_line(conf, buffer, BUF_SIZE) > 0) {
+      D(("Read line: %s", buffer));
+      lines = realloc(lines, ++i * sizeof(char**));
+      if (lines == NULL) {
+	pam_syslog(pamh, LOG_ERR, "Cannot allocate memory.");
+	(void) fclose(conf);
+	return NULL;
+      }
+      lines[i-1] = strdup(buffer);
+    }
+    /* realloc one extra element for the last 0 */
+    lines = realloc(lines, ++i * sizeof(char**));
```

Comment:
Every memory allocation function can fail any time.

But why reading the whole file into memory in the first place?